template/contracts/src/WhitelistNFT.sol

```diff
@@ -0,0 +1,202 @@
+        whitelistClose = _whitelistClose;
+        _baseURIString = _uri;
+        maxWhitelistMint = _maxWhitelistMint;
+        maxPublicMint = _maxPublicMint;
+    }
+
+    /**
+     * @dev Function to mint NFTs from the whitelist
+     * @param _proof Merkle proof of the address
+     * @param _amount Amount of NFTs to mint
+     */
+    function whitelistMint(bytes32[] calldata _proof, uint256 _amount) external payable {
+        if (!live) revert MintNotLive();
+        if (block.timestamp < whitelistOpen) revert WhitelistNotLive();
+        if (block.timestamp > whitelistClose) revert WhitelistNotLive();
+        uint256 minted = _numberMinted(msg.sender) + _amount;
+        if (minted > maxWhitelistMint) revert MintExceeded();
+        if (_totalMinted() + _amount > MAX_SUPPLY) revert SupplyExceeded();
+        if (!MerkleProofLib.verifyCalldata(_proof, whitelistRoot, keccak256(abi.encodePacked(msg.sender)))) {
+            revert WhitelistMintUnauthorized();
+        }
+        if (msg.value != _amount * whitelistPrice) revert InsufficientPayment();
+
+        _mint(msg.sender, _amount);
+    }
+
+    /**
+     * @dev Function to mint NFTs from the public mint
+     * @param _amount Amount of NFTs to mint
+     */
+    function publicMint(uint256 _amount) external payable {
+        if (!live) revert MintNotLive();
+        if (block.timestamp < whitelistClose) revert PublicMintNotLive();
+        uint256 minted = _numberMinted(msg.sender) + _amount;
+        if (minted > maxPublicMint) revert MintExceeded();
+        if (_totalMinted() + _amount > MAX_SUPPLY) revert SupplyExceeded();
+        if (msg.value != _amount * price) revert InsufficientPayment();
+
+        _mint(msg.sender, _amount);
+    }
+
+    /**
+     * @dev Function to set the prices of each NFT
+     * @dev Only the owner can call this function
+     * @param _price Price of each NFT
+     * @param _whitelistPrice Price of each whitelist NFT
+     */
+    function setPrices(uint120 _price, uint120 _whitelistPrice) external onlyOwner {
+        price = _price;
+        whitelistPrice = _whitelistPrice;
+    }
+
+    /**
+     * @dev Function to toggle the minting to live or not
+     */
+    function toggleLive() external onlyOwner {
+        live = !live;
+    }
+
+    /**
+     * @dev Function to set the whitelist root
+     * @param _whitelistRoot Merkle root of the whitelist
+     */
+    function setWhitelistRoot(bytes32 _whitelistRoot) external onlyOwner {
+        whitelistRoot = _whitelistRoot;
+    }
+
+    /**
+     * @dev Function to set the whitelist minting window
+     * @param _whitelistOpen Timestamp of when the whitelist opens
+     * @param _whitelistClose Timestamp of when the whitelist closes
+     */
+    function setWhitelistMintWindow(uint64 _whitelistOpen, uint64 _whitelistClose) external onlyOwner {
+        if (_whitelistOpen > _whitelistClose) revert InvalidWhitelistWindow();
+        if (_whitelistOpen == 0) revert InvalidWhitelistWindow();
```

Comment:
nit: Wouldnt it still be invalid if `_whitelistClose == 0`?

Reply:
Yeah this won't allow `whitelistClose` to be 0.

But in this case essentially you don't really want to use the whitelist mint functionality (since `block.timestamp will always be > 0)`, so you'd be better off using a contract that doesn't implement this functionality since it would just be increasing deploy bytecode size and increasing deploy cost unnecessarily.

Reply:
Okay sounds good!